pageserver/src/tenant.rs

```diff
@@ -5524,27 +5531,27 @@ impl Tenant {
             .unwrap_or(0)
     }
 
-    /// Serialize and write the latest TenantManifest to remote storage.
-    pub(crate) async fn store_tenant_manifest(&self) -> Result<(), TenantManifestError> {
-        // Only one manifest write may be done at at time, and the contents of the manifest
-        // must be loaded while holding this lock. This makes it safe to call this function
-        // from anywhere without worrying about colliding updates.
+    /// Builds a new tenant manifest, and uploads it if it differs from the last-known tenant
+    /// manifest in `Self::tenant_manifest`.
+    pub(crate) async fn maybe_upload_tenant_manifest(&self) -> Result<(), TenantManifestError> {
+        // Serialize manifest updates to avoid races that may clobber it.
         let mut guard = tokio::select! {
-            g = self.tenant_manifest_upload.lock() => {
-                g
-            },
-            _ = self.cancel.cancelled() => {
-                return Err(TenantManifestError::Cancelled);
-            }
+            guard = self.remote_tenant_manifest.lock() => guard,
+            _ = self.cancel.cancelled() => return Err(TenantManifestError::Cancelled),
         };
```

Comment:
Why aren't we using std mutex + try_lock + either unwrap() or bail with "Implementation Error"?

There's no legitimate code path where we would be requesting upload from two places concurrently

Reply:
I correct myself, there are actually concurrent cases:
1. offload
2. unoffload
2. timeline deletion

These can happen simultaneously for different timelines (I haven't checked whether they can happen simultaneously for the same timeline)

Reply:
Also I find it worrisome that it's easy to update runtime state that influences the `build_tenant_manifest` output,
without calling `maybe_upload_tenant_manifest`. The remote timeline client does a better job at this for index part.

Reply:
At a minimm, let's add some more commentary

```suggestion
        // Multiple tasks may be calling this function concurrently after performing mutations to the
        // Tenant object runtime state, resulting in different output from `build_tenant_manifest`.
        // We use an async mutex to serialize these callers.
        // The `eq_ignoring_version` acts as a slightly inefficient but simple coalescing mechanism.
         will may update the Tenant object in a way thThis method is called from multiple tasks
        let mut guard = tokio::select! {
            guard = self.remote_tenant_manifest.lock() => guard,
            _ = self.cancel.cancelled() => return Err(TenantManifestError::Cancelled),
        };
```

Reply:
> The remote timeline client does a better job at this for index part.

That's fair, but unlike the index part, the tenant manifest won't always be the authoritative source of data -- at least not for the stripe size, which is the motivating property here.

I added a TODO to revisit this when we use the manifest for more stuff and have a better grasp of the data model.